### PR TITLE
prov/efa: Bugfix, only disable the cache per domain when checking for FI_MR_LOCAL.

### PR DESCRIPTION
**Pull request by wckzhang**
Due to efa_mr_cache_enable being a global variable, it is checked for
every domain. The check for FI_MR_LOCAL to disable efa_mr_cache_enable
caused a bug where domains with different mr_mode settings would cause
the domains with caches to not cleanup their cache. Due to needing a
domain specific check, we now check if the cache was malloc'd or not and
do not change efa_mr_cache_enable.

### CODE REVIEW

**Comment by wckzhang** [comment]
There's a segfault on RHEL ami's oddly enough:
```
server_cmd: /home/ec2-user/libfabric/fabtests/install/bin/fi_dgram_pingpong -k -p "efa" -E
server_stdout: |
[error] fabtests:common/shared.c:1524: fi_close: Invalid argument(-22) fid 0
bash: line 1: 12335 Segmentation fault /home/ec2-user/libfabric/fabtests/install/bin/fi_dgram_pingpong -k -p "efa" -E
client_cmd: /home/ec2-user/libfabric/fabtests/install/bin/fi_dgram_pingpong -k -p "efa" -E 127.0.0.1
client_stdout: |
[error] fabtests:common/shared.c:1524: fi_close: Invalid argument(-22) fid 0
bash: line 1: 12355 Segmentation fault /home/ec2-user/libfabric/fabtests/install/bin/fi_dgram_pingpong -k -p "efa" -E 127.0.0.1
```

File: prov/efa/src/efa_domain.c [hunk 1]
```diff
@@ -251,7 +245,12 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 		return -FI_EINVAL;
 	}
 
-	if (efa_mr_cache_enable) {
+	/*
+	 * If FI_MR_LOCAL is set, we do not want to use the
+	 * MR cache
+	 */
+	if (efa_mr_cache_enable && info->domain_attr && info->domain_attr->mr_mode & ~FI_MR_LOCAL) {
```
Comment (rwespetal) [inline prov/efa/src/efa_domain.c:252]: This needs to happen before the fork support check, see line 237 in this file

File: prov/efa/src/efa_domain.c [hunk 2]
```diff
@@ -251,7 +245,12 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
...
+		domain->cache = (struct ofi_mr_cache *)malloc(sizeof(struct ofi_mr_cache));
```
Comment (rwespetal) [inline prov/efa/src/efa_domain.c:253]: malloc return needs to be checked

**Review by rwespetal — COMMENTED** [review]
cscope db's were added, please remove them.

Is this thread safe? And is there anywhere else we're changing that environment variable?

File: prov/efa/src/efa_domain.c (hunk 1 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:252]: Wait why does it need to happen before? It won't be malloc'd before fork support so the cache won't be initialized, no?

File: prov/efa/src/efa_domain.c (hunk 2 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:253]: Will do

**Comment by wckzhang** [comment]
> cscope db's were added, please remove them.
 Ah... oops

> Is this thread safe? And is there anywhere else we're changing that environment variable?

It's thread safe, this variable is never changed outside of reading the environment var.

File: prov/efa/src/efa_domain.c (hunk 1 above)
Comment (rwespetal) [reply prov/efa/src/efa_domain.c:252]: `efa_mr_cache_enable` will be 1 (the default) and NCCL will turn on fork support so the earlier check will fail. We'll never get to this if block.

Comment (wckzhang) [reply prov/efa/src/efa_domain.c:252]: hmm alright

**Comment by wckzhang** [comment]
Fixed the segfault...
Called free(domain->cache); but didn't set domain->cache = NULL; after.

**Comment by wckzhang** [comment]
Digging deeper, the malloc call was setting a type which caused an EINVAL in rhel. Switched to calloc, but now I'm seeing: 

```
*** Error in `/home/ec2-user/deferfab/bin/fi_dgram_pingpong': malloc(): memory corruption: 0x0000000002608b60 ***
```

Going to keep digging

**Comment by wckzhang** [comment]
Fixed, misused calloc, should pass everything now

**Comment by zhngaj** [comment]
aws:bot:retest

File: prov/efa/src/efa.h [hunk 3]
```diff
@@ -469,4 +469,9 @@ static inline bool efa_ep_is_cuda_mr(struct efa_mr *efa_mr)
 	return efa_mr ? (efa_mr->peer.iface == FI_HMEM_CUDA): false;
 }
 
+static inline bool efa_is_cache_available(struct efa_domain *efa_domain)
+{
+	return efa_domain->cache ? efa_mr_cache_enable : false;
+}
```
Comment (zhngaj) [inline prov/efa/src/efa.h:475]: The return value is correct. I think `efa_mr_cache_enable` is guaranteed to be 1/true, if efa_domain->cache is not NULL though.

File: prov/efa/src/efa.h [hunk 4]
```diff
@@ -469,4 +469,9 @@ static inline bool efa_ep_is_cuda_mr(struct efa_mr *efa_mr)
 	return efa_mr ? (efa_mr->peer.iface == FI_HMEM_CUDA): false;
 }
 
+static inline bool efa_is_cache_available(struct efa_domain *efa_domain)
+{
+	return efa_domain->cache ? efa_mr_cache_enable : false;
```
Comment (rwespetal) [inline prov/efa/src/efa.h:474]: This check doesn't seem right. As it's written this could return false if `efa_mr_cache_enable` is 0 but `efa_domain->cache` is not NULL. I realize that might not happen but why make it more complicated than it needs to be?

Shouldn't we check `efa_mr_cache_enable` before cache init, and then use `efa_domain->cache` for all other checks in the future? Whether `efa_domain->cache` is set should be the source of truth for whether that domain has a cache, not a global variable.

File: prov/efa/src/efa_domain.c [hunk 5]
```diff
@@ -236,22 +239,27 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
...
 	 */
-	if (info->domain_attr && info->domain_attr->mr_mode & FI_MR_LOCAL) {
-		efa_mr_cache_enable = 0;
+	if (efa_mr_cache_enable && info->domain_attr && info->domain_attr->mr_mode & ~FI_MR_LOCAL) {
+		domain->cache = (struct ofi_mr_cache *)calloc(1, sizeof(struct ofi_mr_cache));
+		if (!domain->cache) {
+			ret = -FI_ENOMEM;
+			goto err_free_info;
+		}
 	}
```
Comment (rwespetal) [inline prov/efa/src/efa_domain.c:248]: I suspect calloc is being called here but I'd be explicit anyway and set domain->cache to NULL when not enabled.

File: prov/efa/src/efa_domain.c [hunk 6]
```diff
@@ -273,9 +281,12 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 		}
 	}
 
-	efa_mr_cache_enable = 0;
-
+	free(domain->cache);
```
Comment (rwespetal) [inline prov/efa/src/efa_domain.c:284]: I think this should be refactored, let's just not allocate this if there is not going to be a cache.

**Review by rwespetal — CHANGES_REQUESTED** [review]

File: prov/efa/src/efa.h (hunk 4 above)
Comment (wckzhang) [reply prov/efa/src/efa.h:474]: If efa_mr_cache_enable is 0 then we want it to return though, 'cause we specified that we don't want a cache.

File: prov/efa/src/efa_domain.c (hunk 5 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:248]: Yeah I can do that

File: prov/efa/src/efa.h (hunk 4 above)
Comment (rwespetal) [reply prov/efa/src/efa.h:474]: But we allocated a cache? That makes no sense.

File: prov/efa/src/efa_domain.c (hunk 6 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:284]: Doesn't work that way, we need to allocate before we call ofi_mr_cache_init. If the init fails, we need to free the cache.

File: prov/efa/src/efa.h (hunk 4 above)
Comment (wckzhang) [reply prov/efa/src/efa.h:474]: I guess it could be:
```
return efa_mr_cache_enable ? efa_domain->cache : false;
```
or
```
return efa_domain->cache ? true : false;
```
It ends up doing the same thing though.

If the cache is not available, return false. If the cache is available return true (efa_mr_cache_enable should be true at this point).

Comment (rwespetal) [reply prov/efa/src/efa.h:474]: I'm failing to see why we need to look at `efa_mr_cache_enable` at all after domain init.

Either `domain->cache` is not NULL and we have a cache, or it's NULL and we don't have a cache.

File: prov/efa/src/efa_domain.c (hunk 6 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:285]: I refactored it a tad, added an int which is checked, could be a bool but not that important.

File: prov/efa/src/efa.h (hunk 4 above)
Comment (wckzhang) [reply prov/efa/src/efa.h]: Yeah I changed it to just ```return efa_domain->cache;```

I'm going to leave it as an in-line so it's easier to understand.

File: prov/efa/src/efa_domain.c (hunk 6 above)
Comment (rwespetal) [reply prov/efa/src/efa_domain.c:285]: Oh, right, forgot about the fall through here. A comment might be helpful here but not required for this PR.

**Review by rwespetal — APPROVED** [review]

File: prov/efa/src/efa_domain.c [hunk 7]
```diff
@@ -232,26 +235,32 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
...
 	 * If FI_MR_LOCAL is set, we do not want to use the
 	 * MR cache
 	 */
-	if (info->domain_attr && info->domain_attr->mr_mode & FI_MR_LOCAL) {
-		efa_mr_cache_enable = 0;
-	}
+	enable_cache = efa_mr_cache_enable && info->domain_attr && info->domain_attr->mr_mode & ~FI_MR_LOCAL;
 
-	if (efa_mr_cache_enable && ofi_vrb_check_fork_enabled(*domain_fid)) {
+	if (enable_cache && ofi_vrb_check_fork_enabled(*domain_fid)) {
```
Comment (shefty) [inline prov/efa/src/efa_domain.c:245]: Posting comment here as well.  The PR that added ofi_vrb_check_fork_enabled() results in a runtime error, I believe because libibverbs isn't linked.

https://github.com/ofiwg/libfabric/issues/6327

For now, the quick fix is to move that code back into the EFA provider (since verbs isn't using it at the moment).  We can figure out where to put shared verbs related code after v1.11.1 is released.

**Comment by rwespetal** [comment]
This PR (and the `FI_MR_LOCAL` check that was merged earlier) is not looking at the right info object so this needs to be reworked.

File: prov/efa/src/efa_domain.c [hunk 8]
```diff
@@ -245,10 +249,18 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 		goto err_close_domain;
 	}
 
-	if (EFA_EP_TYPE_IS_RDM(info))
+	if (EFA_EP_TYPE_IS_RDM(info)) {
 		domain->type = EFA_DOMAIN_RDM;
-	else
+		rxr_domain = container_of(domain_fid, struct rxr_domain,
```
Comment (rwespetal) [inline prov/efa/src/efa_domain.c:254]: Had to stare at this longer than I'd care to admit but this is the right pointer math.

**Review by rwespetal — APPROVED** [review]

File: prov/efa/src/efa_domain.c (hunk 8 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:254]: Not going to lie I stared at this for more than half an hour but it compiled and worked on my first try so yay \o/

File: prov/efa/src/efa_domain.c [hunk 9]
```diff
@@ -301,11 +303,11 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 			                         EFA_MR_CACHE_LIMIT_MULT;
 		cache_params.max_cnt = efa_mr_max_cached_count;
 		cache_params.max_size = efa_mr_max_cached_size;
-		domain->cache.entry_data_size = sizeof(struct efa_mr);
-		domain->cache.add_region = efa_mr_cache_entry_reg;
-		domain->cache.delete_region = efa_mr_cache_entry_dereg;
+		domain->cache->entry_data_size = sizeof(struct efa_mr);
```
Comment (bwbarrett) [inline prov/efa/src/efa_domain.c:317]: This commit is not correct; cache has no backing when you write into it.  I think you fixed this in a later commit, but we should strive to have all commits functional, to not break bisectability.

File: prov/efa/src/efa.h [hunk 10]
```diff
@@ -469,4 +469,9 @@ static inline bool efa_ep_is_cuda_mr(struct efa_mr *efa_mr)
 	return efa_mr ? (efa_mr->peer.iface == FI_HMEM_CUDA): false;
 }
 
+static inline bool efa_is_cache_available(struct efa_domain *efa_domain)
```
Comment (bwbarrett) [inline prov/efa/src/efa.h:472]: Documentation?

File: prov/efa/src/efa_domain.c [hunk 11]
```diff
@@ -206,9 +207,10 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 {
 	struct efa_domain *domain;
 	struct efa_fabric *fabric;
+	struct rxr_domain *rxr_domain;
 	const struct fi_info *fi;
 	size_t qp_table_size;
-	int ret;
+	int ret, enable_cache = 0;
```
Comment (bwbarrett) [inline prov/efa/src/efa_domain.c:213]: Libfabric appears to require the bool type; we should use it for `enable_cache`.

File: prov/efa/src/efa_domain.c [hunk 12]
```diff
@@ -206,9 +207,10 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 {
 	struct efa_domain *domain;
 	struct efa_fabric *fabric;
+	struct rxr_domain *rxr_domain;
```
Comment (bwbarrett) [inline prov/efa/src/efa_domain.c:210]: Why declare this at top of function scope, when it is only used in a single code block?

File: prov/efa/src/efa_domain.c [hunk 13]
```diff
@@ -316,8 +327,8 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 		}
 	}
 
-	efa_mr_cache_enable = 0;
-
+	free(domain->cache);
```
Comment (bwbarrett) [inline prov/efa/src/efa_domain.c:330]: Oh, God, this makes my eyes bleed.  I know this wasn't your fault originally, but we shouldn't write code that is hard to follow.  The primary control flow should be the happy case, with jumps to the error cases.  The happy case being buried in an if statement is awful.  Please flip line 322's logic to jump on error, and refactor so it is obvious this code is in the error path.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: prov/efa/src/efa_domain.c (hunk 11 above)
Comment (rwespetal) [reply prov/efa/src/efa_domain.c:213]: I have to make a change in a future PR here so I can resolve this.

File: prov/efa/src/efa_domain.c (hunk 9 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:317]: Ah yeah my bad, my hunks weren't splittable so I manually deleted and re-added code, must have messed up here.

File: prov/efa/src/efa_domain.c (hunk 13 above)
Comment (wckzhang) [reply prov/efa/src/efa_domain.c:330]: It's not exactly an error path, since it will proceed without the cache...I'll try my best
